src/scalar/mod.rs

```diff
@@ -144,7 +146,22 @@ pub fn new_scalar(array: &dyn Array, index: usize) -> Box<dyn Scalar> {
             };
             Box::new(FixedSizeListScalar::new(array.data_type().clone(), value))
         }
-        Union | Map => todo!(),
+        Union => {
+            let array = array.as_any().downcast_ref::<UnionArray>().unwrap();
+            let type_ = array.types()[index];
+            let (field_index, index) = array.index(index);
+            let field_value = new_scalar(&*array.fields()[field_index], index);
+            Box::new(UnionScalar::new(
+                array.data_type().clone(),
+                type_,
+                if field_value.is_valid() {
+                    Some(field_value.into())
+                } else {
+                    None
+                },
+            ))
+        }
```

Comment:
```suggestion
        Union => {
            let array = array.as_any().downcast_ref::<UnionArray>().unwrap();
            Box::new(UnionScalar::new(
                array.data_type().clone(),
                array.value(index),
            ))
        }
```

This would be my suggestion